### DISCUSSION TITLE
Extracting a value from an IO[Boolean]

### DISCUSSION THREAD

**Discussion by LeeTibbert** [armanbilge/epollcat#57] [General]
@armanbilge 

Re: Issue #52, checking for (only) IPv4 on epollcat startup.

I think the time has come to ask for help.  I have spent most of the
day and am beyond the zone of learning.  Thank you.

What I am trying to do is, given an IO[Boolean], extract/materialize
the Boolean value.  You probably know the idiom without hesitation.
I will post the (eyesore) code below, so you can see where I am
leaving the rails. 

I believe that epollcat is using Cats Effect 3, a snapshot/early-build
of 3.3.  I am using the documentation. I am using the CE 3
documentation at https://typelevel.org/cats-effect/api/3.x/cats/effect/IO.html
& related.

The answers on the Web are all of the sort "Don't even try, just pass the IO along"

The documentation says that I should be doing something like 
"val program = IO(mumble) program.unsafeRunSync()"
The compiler complains that `unsafeRunSync is not a member of IO[Boolean]'
and suggests `unsafeRunAsync`. Fair enough.

`unsafeRunAsync` requires a callback, which I could probably figure out.
Having the Boolean in a callback does me no good, I need it before
the rest of the test Suite proceeds.

Down to code.   At the beginning of `IPv6Suite.scala`

```
object NetworkProtocolInfo {
 def useIPv4Stack(): Boolean = {
    val program =
    IOServerSocketChannel
      .open
      .use { ch =>
        for {
          sz <- IO(ch.localAddress
          .asInstanceOf[InetSocketAddress]
          .getAddress()
          .getAddress()
          .size)
          _ <- IO(printf(s"sz (sin size): ${sz}\n\n"))
          // Fixme - Check if either exactly 4 or exactly 16, or else fail.     
          result <- IO(sz == 4)
          _ <- IO(printf(s"result: ${result}\n\n"))
        } yield result  
      } // At this point, the compiler tells me I have an IO[Boolean], understood

     //    program.unsafeRunSync()  // Compiler rejects with uRunSync() "is not a member of IO[Boolean]]                                                         
     // program.unsafeRunAsync(/* callback */) // I need value inside IO[Boolean], when run, at this point, not in callback

     true // FIXME -- DEBUG scaffolding                                        
  }
} // NetworkProtocolInfo 


```
When I use the snippet inside the "program" as part of a 
test, it all works, or did in some variant.

Thank you.

**Comment by armanbilge** [comment]
> The answers on the Web are all of the sort "Don't even try, just pass the IO along"

Yes, I'm afraid this is the correct strategy 😅 In particular `unsafeRunSync()` is not available because Scala Native is single-threaded.

Can you write it as `IO(assertEquals(sz, 4))`? Or will you need the boolean for some other purpose?

**Reply by armanbilge** [reply]
Oh, perhaps I misunderstood. Is this for the test suite, or for the main code?

**Reply by LeeTibbert** [reply]
This is for pre-amble code which runs before the test suites.
At the top of each Suite
```
override def munitIOTimeout = 10.seconds

  override def munitIgnore: Boolean =
    NetworkProtocolInfo.useIPv4Stack() == false

  test("Bind ::1 - No IPv6") { // example
}
} // End of Suite
```
You/we had talked/written the other day about dealing with the case where if only IPv4
is available, one set of tests would intercept the expected "AF family not supported" (not exact words)
exception.

If the Suite "Knows" it is safely on IPv6 it can proceed to look for defects.

I can (and have) punted this to manually setting the value, but am trying to 
push the peg about and automate it, so that others can follow and do
not have to do manual futzing.

**Reply by armanbilge** [reply]
I see. I think the easiest thing to do here is just use `AsynchronousServerSocketChannel` directly instead of `IOServerSocketChannel` to determine the value. The `IOServerSocketChannel` are designed to make testing concurrent sockets easier, but we don't have to use them for everything.

**Comment by LeeTibbert** [comment]
I can not use a Scala Socket and the same trick to "peek" into it, because SN 0.4.0 will, modulo bugs,
always return IPv4, and epollcat wants to use IPv6.

I will be documenting in SocketHelpers.scala that epollcat can & should use the 'peek into the SN socket and 
see what it is using' to initialize the `useIPv4Stack` (née preferIPv4Stack).

**Reply by LeeTibbert** [reply]
> `AsynchronousServerSocketChannel` 

I will try that, thank you.

Where did epollcat and/or typelevel come up with the 
`maxColumn = 96`?  in `.scalafmt.conf`?  Seems like
an odd (yet even) value.  Now I know why my
lines do not get wrapped at 80 columns.

**Reply by armanbilge** [reply]
I copied it from Cats Effect. Feel free to change it if you want; I don't have strong preferences. So long as the code is tidy I'm happy :)
https://github.com/typelevel/cats-effect/blob/3523728a6b154878bb4425549b161b5f38800891/.scalafmt.conf#L14

**Reply by LeeTibbert** [reply]
OK, if it does not throw epollcat into out-of-sync. I will submit an independent PR to change it
to the linux standare 80 (79?) and you can consider it at your own pace, not as an interruption.

**Comment by LeeTibbert** [comment]
> use AsynchronousServerSocketChannel directly 

That was the info I needed to get me unstuck.  I have a first
evolution of Tcp6Suite working which checks  expected 
Exception if epollcat is using IPv4 and does IPv6 tests if
it is using IPv6.   

This gives fewer surprises to someone who clones the Repository
and runs the tests, neither knowing nor caring about which IP they
are using.

**Comment by LeeTibbert** [comment]
re: formatting

I gave it a go in my private workspace and backed off.  To many gratuitous changes for 
existing code.  Not all ideas/desires get past infancy.  I want to work on epollcat & network
stuff, not fight scalafmt.

> So long as the code is tidy I'm happy :)

I'll sing in that choir...

**Comment by LeeTibbert** [comment]
I have the Tcp6Suite testing environment working as a first evolution and sensitive to IPv4 vs IPv6.
It is currently a few pinpoint but distinguishing tests. 

Next sprint I will work on implementing the code in SocketHelpers which it, indirectly, tests.
